reference/best-practices: add PD scheduling best practice (#1655)
pingcap/docs: dev/reference/best-practices/pd-scheduling.md

---
title: PD Scheduling
summary: Learn best practice and strategy for PD scheduling.
category: reference
---

# PD Scheduling

This document details the principles and strategies of PD scheduling through common scenarios to facilitate user application. This document assumes that you have a basic understanding of TiDB, TiKV and PD with the following core concepts:

- [Leader/Follower/Learner](/dev/glossary.md#leaderfollowerlearner)
- [Operator](/dev/glossary.md#operator)
- [Operator Step](/dev/glossary.md#operator-step)
- [Pending/Down](/dev/glossary.md#pendingdown)
- [Region/Peer/Raft Group](/dev/glossary.md#regionpeerraft-group)
- [Region Split](/dev/glossary.md#region-split)
- [Scheduler](/dev/glossary.md#scheduler)
- [Store](/dev/glossary.md#store)

> **Note:**
>
> This document initially targets TiDB 3.0. Though some functions are not supported in earlier versions (2.x), but this document still can be used as a reference for similar principles.

## PD scheduling policies

This section introduces the principle and process involved in the scheduling system.

### Scheduling process

> Scheduling process generally has three steps:

Reviewer: ```suggestion
The scheduling process generally has three steps:
```

1. Collect information

>     Each TiKV node periodically reports two heartbeats to PD: `StoreHeartbeat` and `RegionHeartbeat`. `StoreHeartbeat` contains the overall information of Stores including disk capacity, remaining storage, reads and writes traffic. `RegionHeartbeat` contains the overall information of Regions including the range of each Region, peer distribution, peer status, data volume, and reads and writes traffic. PD collects and restores these information for scheduling decisions.

Reviewer 2: two heartbeats -> two types of heartbeats

Reviewer 2: `and reads and writes traffic` -> `and read/write traffic`?

>     Each TiKV node periodically reports two types of heartbeats to PD: `StoreHeartbeat` and `RegionHeartbeat`. `StoreHeartbeat` contains the overall information of Stores including disk capacity, remaining storage, read/write traffic. `RegionHeartbeat` contains the overall information of Regions including the range of each Region, peer distribution, peer status, data volume, and reads and writes traffic. PD collects and restores these information for scheduling decisions.

Reviewer: ```suggestion
    Each TiKV node periodically reports two types of heartbeats to PD: `StoreHeartbeat` and `RegionHeartbeat`. `StoreHeartbeat` contains the overall information of Stores including disk capacity, remaining storage, read/write traffic. `RegionHeartbeat` contains the overall information of Regions including the range of each Region, peer distribution, peer status, data volume, and reads and writes traffic. PD collects and stores this information for scheduling decisions.
```

2. Generate Operators

    Different schedulers generate Operators to be executed based on their own logic and requirements, with the consideration of constraints and limitations include:

>      - Do not add Peers to a Store with abnormal states (disconnect, down, busy, out of space, with extensive sent/received Snapshots)

Reviewer: I think "sending/receiving snapshots intensely" counts as "Busy" no?

     - Do not balance Regions with abnormal states
     - Do not attempt to transfer a Leader to a Pending Peer
     - Do not attempt to remove a Leader directly
     - Do not break the physical isolation of various Region Peers
     - Do not break constraints such as Label property

3. Execute scheduling tasks

    The generated Operator first joins a queue managed by `OperatorController` rather than be executed immediately. The `OperatorController` takes the Operator out of the queue and executes it with a certain amount of concurrency based on the configuration. The procedure is to distribute each Operator Step to the corresponding Region Leader. In the end, the Operator is marked as finish or timeout and removed from the execution list.

### Load balancing

> Region primarily relies on `balance-leader` and `balance-region` schedulers to achieve load balance. Both schedulers target distributing Regions evenly across all Stores in the cluster but with separate focuses: `balance-leader` attends Region Leader to balance incoming client requests. `balance-region` centers each Region Peer to redistribute the pressure of storage while avoid exceptions like storage failure.

Reviewer 2: `exceptions like storage failure` -> `exceptions like storage out of space`

> Region primarily relies on `balance-leader` and `balance-region` schedulers to achieve load balance. Both schedulers target distributing Regions evenly across all Stores in the cluster but with separate focuses: `balance-leader` attends Region Leader to balance incoming client requests. `balance-region` centers each Region Peer to redistribute the pressure of storage while avoid exceptions like out of storage space.

Reviewer: I think the 'attends' in this sounds a bit awkward:

>  `balance-leader` **attends** Region Leader to balance incoming client requests.

Similarly 'centers each' also sounds quite strange in:

>  `balance-region` **centers each** Region Peer to redistribute the pressure of storage while avoid exceptions like out of storage space.

Maybe there is a different way we can work them to make the meaning more clear?

Author: Thank you, Ana. The intention here is to say:
"`balance-leader` focuses on Region Leaders while `balance-region` focuses on each Region Peer..." But I don't want to overuse "focus" in one paragraph.
How about "deal with"?
"`balance-leader` deals with Region Leaders to balance incoming client requests, whereas `balance-region` deals with each Region Peer to redistribute the pressure of storage and avoid exceptions like out of storage space."

Reviewer: I think this is a great improvement!

Reviewer: If you want to avoid repeating "Deal with" you could do like:

> `balance-leader` deals with Region Leaders to balance incoming client requests, whereas `balance-region` concerns itself with each Region Peer to redistribute the pressure of storage and avoid exceptions like out of storage space.

### Cluster topology awareness

Cluster topology awareness (zone/rack/host awareness) is having the knowledge of cluster topology or more specifically how the different data nodes are distributed. This is to make the different Regions Peers as distributed as possible through scheduling, hence to achieve high availability and disaster recovery. PD continuously scans all Regions in the background. When PD finds that the distribution of Regions is not optimal, it generates a Operator to replace Peers and redistribute Regions.

The component to check the distribution of Regions is `replicaChecker`, which is similar to Scheduler except that it cannot be disabled. `replicaChecker` checks and schedules based on the information provided by the configuration item `location-labels`. For example, `[zone, rack, host]` defines a three-tier topology: the cluster is configured with multiple available zones, with multiple racks under each zone and multiple hosts under each rack. PD attempts to schedule Region Peers to different zones first, or to different racks when zones are limited, or to different hosts when racks are limited.

### Scale-down and failure recovery

> Scale-down refers to taking a Store offline. You can use commands to mark the Store as `Offline` while PD replicates the data the offline node held onto other nodes by scheduling. Failure recovery applies when Stores failed and cannot be recovered. In such case, a Region with Peers distributed on the corresponding Store might lose replicas, which requires PD to replenish replicas for these Regions on other nodes.

Reviewer 3: ```suggestion
Scale-down refers to the process when you take a Store offline and mark it as "Offline" using a command, PD replicates the Regions on the offline node to other nodes by scheduling. Failure recovery applies when Stores failed and cannot be recovered. In this case, Regions with Peers distributed on the corresponding Store might lose replicas, which requires PD to replenish on other nodes.
```

### Region health

**Grafana PD/Cluster/Region health** panel shows the statistics about Regions in abnormal states, include Pending Peer, Down Peer, Offline Peer and Regions with extra or few Peers.

You can query the list of Regions in abnormal conditions using pd-ctl with region check commands:

- `region check miss-peer`: Regions without enough Peers
- `region check extra-peer`: Regions with extra Peers
- `region check down-peer`: Regions with Down Peers
- `region check pending-peer`: Regions with Pending Peers

## Scheduling strategy control

You can use pd-ctl to adjust the scheduling strategy from the following three aspects. Refer to [PD Control](/dev/reference/tools/pd-control.md) for more details.

### Start-stop scheduler

pd-ctl supports dynamically creating and deleting Schedulers. You can use the following commands to control the scheduling behavior of PD:

- `scheduler show`: show currently running Schedulers in the system
- `scheduler remove balance-leader-scheduler`: delete (disable) balance-leader-scheduler
- `scheduler add evict-leader-scheduler-1`: add a scheduler to remove all Leaders in Store 1

### Add Operators manually

> Pd also supports creating or removing Schedulers directly through pd-ctl. For example:

Reviewer 2: creating or removing **Operators** directly

- `operator add add-peer 2 5`: add Peers to Region 2 in Store 5
- `operator add transfer-leader 2 5`: migrate Region 2 Leader to Store 5
- `operator add split-region 2`: split Region 2 into two Regions evenly in size
- `operator remove 2`: remove currently pending Operator in Region 2

### Adjust scheduling parameter

> You can check the scheduling configuration using pd-ctl with `config show` command, and adjust the value using `config set {key} {value}`. Common adjustments include:

Reviewer 3: ```suggestion
You can check the scheduling configuration using the config show` command in pd-ctl, and adjust the values using `config set {key} {value}`. Common adjustments include:
```

> - `leader-schedule-limit`: control the number of concurrency of Transfer Leader scheduling

Reviewer 3: ```suggestion
- `leader-schedule-limit`: controls the concurrency of Transfer Leader scheduling
```

Reviewer 3: Please modify the below instances similarly

- `region-schedule-limit`: control the number of concurrency of adding/rdeleting Peer scheduling

> - `disable-replace-offline-replia`: stop taking nodes offline

Reviewer 3: ```suggestion
- `disable-replace-offline-replia`: determines whether to disable the scheduler to take nodes offline
```

Reviewer 3: @disksing replia or replica. Please help check, thanks.

> - `disable-location-replacement`: stop adjusting the isolation level of Regions

Reviewer 3: ```suggestion
- `disable-location-replacement`: determines whether to disable the scheduling that handles the isolation level of Regions
```

> - `max-snapshot-count`: control the maximum of sent/received Snapshots concurrently of each Store

Reviewer 3: ```suggestion
- `max-snapshot-count`: controls the maximum concurrency of sent/received Snapshots for each Store
```

## PD scheduling in common scenarios

This section illustrates the best practice of PD scheduling strategies through several scenarios and their scheduling plans.

### Leader/Region is not evenly distributed

> The grading mechanism of PD determines that Leader Count and Region Count of different Stores cannot fully explain the load balance status. Therefore, it is necessary to confirm whether there is load imbalance from the actual load of TiKV or Storage usage.

Reviewer 3: ```suggestion
The grading mechanism of PD determines that Leader Count and Region Count of different Stores cannot fully reflect the load balancing status. Therefore, it is necessary to confirm whether there is load imbalancing from the actual load of TiKV or Storage usage.
```

Once you have confirmed that Leader/Region is not evenly distributed, you need to check the grading of different Stores.

If the scores of different Stores are close, it means PD mistakenly believes that Leader/Region is evenly distributed. Possible reasons are:

> - There are hotspots which cause load imbalance. In such case, you need to collect information about hot Regions scheduling before taking the next step. For more details, refer to [hotspot scheduling](#hot-regions-is-not-evenly-distributed) below.

Reviewer 3: ```suggestion
- There are hotspots that cause load imbalancing. In such case, you need to analyze further based on [hotspot scheduling](#hot-regions-is-not-evenly-distributed) below.
```

> - There are a large number of empty Regions or small Regions, which leads to a great difference in the number of Leaders in different Stores and further burdens raftstore. This is the time for a Region Merge and quicken merging process. For more details, refer to the [Region Merge](#the-speed-of-region-merge-is-slow) section below.

Reviewer 3: ```suggestion
- There are a large number of empty Regions or small Regions, which leads to a great difference in the number of Leaders in different Stores and high pressure on raftstore. This is the time for a [Region Merge](#the-speed-of-region-merge-is-slow) scheduling.
```

> - Hardware and software environment varies from Store to Store. You can accordingly adjust the value of `leader-weight` and `region-weight` to control the distribution of Leader/Region.

Reviewer 3: ```suggestion
- Hardware and software environment varies among Stores. You can adjust the value of `leader-weight` and `region-weight` accordingly to control the distribution of Leader/Region.
```

You can use pd-ctl to adjust the scheduling strategy from the following three aspects. Refer to [PD Control](/dev/reference/tools/pd-control.md) for more details.

### Start-stop scheduler

pd-ctl supports dynamically creating and deleting Schedulers. You can use the following commands to control the scheduling behavior of PD:

- `scheduler show`: shows currently running Schedulers in the system
- `scheduler remove balance-leader-scheduler`: removes (disable) balance-leader-scheduler
- `scheduler add evict-leader-scheduler-1`: adds a scheduler to remove all Leaders in Store 1

### Add Operators manually

Pd also supports creating or removing Operators directly through pd-ctl. For example:

- `operator add add-peer 2 5`: adds Peers to Region 2 in Store 5
- `operator add transfer-leader 2 5`: migrates Region 2 Leader to Store 5
- `operator add split-region 2`: splits Region 2 into two Regions evenly in size
- `operator remove 2`: removes currently pending Operator in Region 2

### Adjust scheduling parameter

You can check the scheduling configuration using the `config show` command in pd-ctl, and adjust the values using `config set {key} {value}`. Common adjustments include:

> - `leader-schedule-limit`: controls the concurrency of Transfer Leader scheduling

Reviewer 3: ```suggestion
- `leader-schedule-limit`: Controls the concurrency of Transfer Leader scheduling
```
Initial capitalized for all similar instances.

- The schedule speed is limited by default. You can adjust `leader-schedule-limit` or `replica-schedule-limit` to a larger value. Similarly, `max-pending-peer-count` and `max-snapshot-count` can also be properly enlarged.
- Other scheduling tasks are running concurrently and competing in the system. You can refer to the solution in [the previous section](#leaderregion-is-not-evenly-distributed).
- When you take a single node offline, a number of Region Leaders to be operated are concentrated on the offline node (about 1/3 under the configuration of 3 replicas), so the speed is limited by the speed at which this single node generates Snapshots. You can speed it up by manually adding an `evict-leader-scheduler` to migrate Leaders.

If the corresponding Operator fails to generate, possible reasons are:

- The Operator is stopped, or `replica-schedule-limit` is set to 0.
- There is no proper node to migrate Regions. For example, if the capacity of nodes that replace the nodes of same Label is larger than 80%, PD will stop scheduling to avoid running out of storage space. In such case, you need to add more nodes or delete some data to free space.

> ### The speed of putting nodes online is slow

Reviewer 3: ```suggestion
### Bringing nodes online is slow
```

> Currently, to take nodes online is scheduled through balance region mechanism, so you can refer to [Leader/Region is not evenly distributed](#leaderregion-is-not-evenly-distributed) for troubleshooting.

Reviewer 3: ```suggestion
Currently, bringing nodes online is scheduled through the balance region mechanism. You can refer to [Leader/Region is not evenly distributed](#leaderregion-is-not-evenly-distributed) for troubleshooting.
```

### Hot Regions are not evenly distributed

> Hot Regions scheduling generally has the following problems:

Reviewer 3: ```suggestion
Hot Regions scheduling issues generally fall into the following categories:
```

Reviewer 3: scheduling having problems sounds bad for the feature...

> - There is a majority of hot Regions, but the scheduling speed cannot keep up with them to redistribute hot Regions in time.

Reviewer 3: ```suggestion
- Hot Regions can be observed via PD metrics, but the scheduling speed cannot keep up to redistribute hot Regions in time.
```

>     **Solution**: adjust `hot-region-schedule-limit` to a larger value, and reduce the limit quota of other schedulers to speed up hot Regions scheduling. Or you can adjust `hot-region-cache-hits-threshold` to a smaller value to make PD sensitive to traffic changes.

Reviewer 3: ```suggestion
    **Solution**: adjust `hot-region-schedule-limit` to a larger value, and reduce the limit quota of other schedulers to speed up hot Regions scheduling. Or you can adjust `hot-region-cache-hits-threshold` to a smaller value to make PD more sensitive to traffic changes.
```

> - A single Region with extensive traffics. For example, to scan a small table extensively is required in the production environment, which can also be detected from PD metrics. Since a single hotspot cannot be resolved by redistributing, you need to manually add a `split-region` Operator to redistribute such a Region.

Reviewer 3: ```suggestion
- Hotspot formed on a single Region, for example, a small table intensively scanned by massive amount of requests。 This can also be detected from PD metrics. Since you cannot actually scatter a single hotspot, you need to manually add a `split-region` Operator to split such a Region.
```

> - The load of some nodes is significantly higher than that of other nodes from TiKV-related metrics, which becomes the bottleneck of the whole system. Currently, PD counts hotspots through traffic analysis. So it is possible that PD fails to identify hotspots in certain scenarios. For example, some Regions have a large number of point-and-check requests, which are not significant in terms of traffic, but high QPS of which leads to bottlenecks in key modules.

Reviewer 3: ```suggestion
- The load of some nodes is significantly higher than that of other nodes from TiKV-related metrics, which becomes the bottleneck of the whole system. Currently, PD counts hotspots through traffic analysis only, so it is possible that PD fails to identify hotspots in certain scenarios. For example, when there are intensive point lookup requests for some Regions, it might not be obvious to detect in traffic, but still the high QPS might lead to bottlenecks in key modules.
```

Reviewer 3: Try not start a sentence with words like "so", "and", etc.

Reviewer 3: point lookup/query 点查

>     **Solutions**: Firstly, locate the table with extensive traffic by examining operational needs, and add a `scatter-range-scheduler` to make all Regions of this table are evenly distributed. TiDB also provides an interface in its HTTP AIP to simplify this operation. Refer to [TiDB HTTP API](https://github.com/pingcap/tidb/blob/master/docs/tidb_http_api.md) for more details.

Reviewer 3: ```suggestion
    **Solutions**: Firstly, locate the table where hot Regions are formed based on the specific business. Then add a `scatter-range-scheduler` scheduler to make all Regions of this table evenly distributed. TiDB also provides an interface in its HTTP AIP to simplify this operation. Refer to [TiDB HTTP API](https://github.com/pingcap/tidb/blob/master/docs/tidb_http_api.md) for more details.
```

> ### The speed of Region Merge is slow

Reviewer 3: ```suggestion
### Region merge is slow
```

### Taking nodes offline is slow

This scenario requires examining the generation and execution of Operators through related metrics.

If Operators are successfully generated but the scheduling process is slow, possible reasons are:

- The scheduling speed is limited by default. You can adjust `leader-schedule-limit` or `replica-schedule-limit` to larger value.s Similarly, you can consider loosening the limits on `max-pending-peer-count` and `max-snapshot-count`.
- Other scheduling tasks are running concurrently and racing for resources in the system. You can refer to the solution in [the previous section](#leadersregions-are-not-evenly-distributed).
- When you take a single node offline, a number of Region Leaders to be processed (around 1/3 under the configuration of 3 replicas) are distributed on the node to remove. Therefore, the speed is limited by the speed at which snapshots are generated by this single node. You can speed it up by manually adding an `evict-leader-scheduler` to migrate Leaders.

If the corresponding Operator fails to generate, possible reasons are:

- The Operator is stopped, or `replica-schedule-limit` is set to "0".

> - There is no proper node for Region migration. For example, if the available capacity size of the replacing nodes (of the same label) is less than 20%, PD will stop scheduling to avoid running out of storage space. In such case, you need to add more nodes or delete some data to free the space.

Reviewer 3: ```suggestion
- There is no proper node for Region migration. For example, if the available capacity size of the replacing node (of the same label) is less than 20%, PD will stop scheduling to avoid running out of storage on that node. In such case, you need to add more nodes or delete some data to free the space.
```
